plugins/ksp/src/org/jetbrains/kotlin/ksp/symbol/impl/kotlin/KSDeclarationImpl.kt

```diff
@@ -56,4 +54,8 @@ abstract class KSDeclarationImpl(ktDeclaration: KtDeclaration) : KSDeclaration {
     override fun toString(): String {
         return this.simpleName.asString()
     }
+
+    override val isLocal: Boolean by lazy {
+        KtPsiUtil.isLocal(ktDeclaration) && (ktDeclaration as? KtParameter)?.hasValOrVar() != true
```

Comment:
The corresponding `parentDeclaration` can still be the constructor, while it is the class in KSDeclaration(Descriptor|Java)Impl.  This can be very confusing.

Reply:
For example, the same class in the form of source and .class give different results.

Reply:
Can you elaborate? This is the implementation for Kotlin source based symbols. For say descriptor based symbols, properties in constructor parameter list is already resolved to be part of the declarations, therefore the parameter you see in constructor list are just value parameters. The reason for Psi for Kotlin symbols to not work is because the underlying Psi for KSPropertyDeclarationParameterImpl is KtParameter.

Reply:
For example, in
```
class C(val v: Int)
```
I'm wondering whether the `parentDeclaration` of v is `C` for descriptor and java and it's `C.constructor` for the kotlin impl?

Reply:
depends on how you view `v` as a parameter in constructor or a property declared in constructor. In Kt impl parent of `v` is C.constructor.

Just want to point out that `isLocal` here in KSDeclarationImpl is to distinguish between a local variable and a property declared in primary constructor parameter list.

```
class C(val v: Int, irrelevant: Int) {
    constructor(val arg: Int) {
        this(arg+1, 123)
    }
}
```
In above code, I can use isLocal to differentiate `v` and `arg`, and because `irrelevant` is not a declaration at all so I won't see it when inspecting declarations.